Deflate and inflate exponents of only some variables

asbuch opened the pull request:
Given an MPolyElem f and a list of some of the variables in f,
deflate(f, vars, shift, defl) returns a polynomial obtained by deflating
only the specified variables. vars can be an array of either generators
or variable indices. shift and defl var arrays of integers, one for each
of the specified variables. inflate(f, vars, shift, defl) is defined
similarly.

Example:
```
using AbstractAlgebra

R, (x,y,z) = PolynomialRing(ZZ, ["x", "y", "z"])

# How can I generate a random multivariate polynomial?
g = (2*x^2 + y*x + 3*y^2*z - 4*x^3 + x*z - 5*z^2)^2

g0 = coeff(g, [y], [0])
g1 = deflate(g - g0, [y], [1], [1])

@assert g == g0 + y * g1 == g0 + inflate(g1, [y], [1], [1])
```

For some code I am trying to write, some functions only use operations like those in the examples. I assume that inflate and deflate is faster than multiplying and dividing with a monomial?

thofma commented:
Thanks. I think it is a good idea to have this functionality.

P.S.: You can do `rand(R, 1:5, 1:3, -10:10)` to generate a random polynomial.

wbhart commented:
I agree, this is a good idea. I would approve this if it had some basic test code to keep code cov happy and if appropriate changes were made to the documentation to cover the new functions.

wbhart commented:
Also, to answer your question @asbuch I would trust that inflation/deflation is fast more readily than I would trust division by a monomial. In theory that case is probably optimised in Flint, but whether it ever got done in AbstractAlgebra I don't recall.

asbuch commented:
Thanks for all the comments! I added some tests to MPoly-test.jl.

Regarding documentation, the original pull request did come with docstrings for the new methods of inflate and deflate. If more than this is required, then please give some more details about what is required and which files need to be edited.

asbuch commented:
I also tried to add tests for inflate/deflate with UnivPoly, but realized that the new methods don't apply to UnivPoly at all because the only thing UnivPoly and MPoly have in common is that they are both rings. I wonder if there ought to be a type between RingElem and PolyElem, MPolyElem, UnivPolyElem?

This is probably related to other small differences between the various types of polynomial rings. For example, it would be natural to allow gen(R,1) if R is a PolyRing, and to allow gen(R,"x") and gens(R,["x"]) if R is a PolyRing or MPolyRing that contains the variable "x". Changes like that would make it easier to switch between implementations of polynomial rings.

wbhart commented:
The three polynomial modules don't have the same interface so if we put something between them and Ring we'd need to document the common interface.

As things are at the moment, implementations for MPoly won't automatically apply to UnivPoly. In fact, if had not even occurred to me that this should be possible. But maybe there is a common interface that could be extracted there which is not going to be inefficient if stuff is built on top of it.

Anyway, as for the docs, they just need to be instantiated in `docs/src/mpolynomial.md file`, in the inflation/deflation section.

wbhart commented:
If you want to test that the docs are instantiated correctly, change into the `docs` directory (not `docs/src`) and run `make.jl` from Julia, e.g.

```
~/julia-1.6.6/bin/julia make.jl
```

It may complain about missing dependencies which you have to install in Julia in the usual way, but this is not usually difficult. You need Julia 1.6.x to build the docs I believe.

tthsqe12 commented:
Almost there.
For consistency, change the method definition for deflate (and the one in the @ doc string) that takes the variables as integer to match the one for inflate:
`deflate(f::AbstractAlgebra.MPolyElem, vars::Vector{Int}, shift::Vector{Int}, defl::Vector{Int})`
Then, copy and paste these into the relevant sections of `docs/src/mpolynomial.md` around line 640:
```
inflate(f::AbstractAlgebra.MPolyElem, vars::Vector{Int}, shift::Vector{Int}, defl::Vector{Int})
inflate(f::T, vars::Vector{T}, shift::Vector{Int}, defl::Vector{Int}) where T <: AbstractAlgebra.MPolyElem
deflate(f::AbstractAlgebra.MPolyElem, vars::Vector{Int}, shift::Vector{Int}, defl::Vector{Int})
deflate(f::T, vars::Vector{T}, shift::Vector{Int}, defl::Vector{Int}) where T <: AbstractAlgebra.MPolyElem
```

asbuch commented:
Thanks for the instructions. I fixed the signature typo in MPoly.jl and added the signatures and an example to docs/src/mpolynomial.md. And I managed to compile the documentation, nice to know how to do that!

tthsqe12 commented:
By the way, your CO2 graphs are a little out of date.

thofma commented:
Thanks Anders!

asbuch commented:
> By the way, your CO2 graphs are a little out of date.

Correct, I should update some time.
